fastestimator/pipeline.py

```diff
@@ -317,146 +282,35 @@ def _test_loader(self, dataset: OpDataset, batch_size: int, shuffle: bool, num_w
-            if size == tf.data.experimental.UNKNOWN_CARDINALITY:
-                self.num_examples[mode] = self.dataloaders[mode].reduce(0, lambda x, _: x + 1)
-            else:
-                # Cardinality is based on batch size, but we want total number of examples
-                self.num_examples[mode] = int(self.num_examples[mode]) * self.batch_sizes[mode]
-
-    def get_modes(self) -> Set[str]:
-        return set(self.dataloaders.keys())
-
-    def get_global_batch_size(self, mode: str, epoch: int) -> int:
-        return self.batch_sizes.get(mode, 0)
-
-    def get_signature_epochs(self, mode: str) -> List[int]:
-        return [0]
-
-    def get_all_output_keys(self) -> Set[str]:
-        return self.output_keys
-
-    def get_num_examples(self, mode: str, epoch: int) -> int:
-        return self.num_examples.get(mode, 0)
-
-    def transform(self, mode: str, epoch: int = 0, dataset: Optional[Dataset] = None) -> Iterator:
-        if dataset is not None:
-            loader = DataLoader(dataset=dataset,
-                                batch_size=self.batch_sizes.get(mode, 1),
-                                num_workers=os.cpu_count(),
-                                drop_last=False)
-        else:
-            loader = self.dataloaders.get(mode, ())
-        return iter(loader)
-
-
-def _infer_batch_size(data) -> int:
-    if isinstance(data, (torch.Tensor, tf.Tensor)):
-        return data.shape[0]
-    if isinstance(data, (list, tuple)):
-        return _infer_batch_size(data[0])
-    if isinstance(data, set):
-        return _infer_batch_size(data.pop())
-    if isinstance(data, dict):
-        return _infer_batch_size(data.popitem()[1])
-    raise ValueError("Could not infer batch size from data loader")
+    def get_iterator(self, mode: str, epoch: int) -> Iterator:
+        return self.dataloaders[mode]
+
+    def get_batch_size(self, mode: str, epoch: int) -> int:
+        dataloader = self.get_iterator(mode=mode, epoch=epoch)
+        return dataloader.batch_size
+
+    def get_num_steps(self, mode: str, epoch: int) -> int:
+        dataloader = self.get_iterator(mode=mode, epoch=epoch)
+        return len(dataloader)
+
+    def index(self, idx: int, mode: str = "train", epoch: int = 0) -> dict:
+        dataloader = self.get_iterator(mode=mode, epoch=epoch)
+        dataset = dataloader.dataset
+        return dataset[idx]
+
+
+def Pipeline(train_data: Union[Dataset, DataLoader, Scheduler, tf.data.Dataset],
+             eval_data: Union[Dataset, DataLoader, Scheduler, tf.data.Dataset, None] = None,
+             batch_size: Union[int, Scheduler, None] = None,
+             ops: Union[None, NumpyOp, Scheduler[NumpyOp], Iterable[Union[NumpyOp, Scheduler[NumpyOp]]]] = None,
+             num_process: Union[int, None] = None):
+    assert isinstance(train_data, (Dataset, DataLoader, Scheduler, tf.data.Dataset))
+    if not eval_data:
+        assert type(train_data) == type(eval_data)
```

Comment:
this check needlessly prevents someone from, for example, having NumpyData for training but GeneratorData for testing